pkg/pillar/cmd/upgradeconverter/upgradeconverter.go

```diff
@@ -0,0 +1,65 @@
+package upgradeconverter
+
+import (
+	"flag"
+	"log"
+)
+
+// ConverterAction - Defines the types of action that can occur when moving between versions
+type ConverterAction uint8
+
+const (
+	// Convert - Convert from old to new
+	Convert ConverterAction = iota + 1
+	// CleanupOld - Delete old versions
+	CleanupOld
+	// CleanupNew - Delete new versions if outdated
+	CleanupNew
+	// NoAction - Do nothing
+	NoAction
+)
+
+var conversionHandlers = []ConversionHandler{
+	{
+		description: "Convert Global Settings to new format",
+		handlerFunc: handleUpgradeGlobalConfig,
+	},
+}
+
+// Run - runs the main upgradeconverter process
+func Run() {
+	actionStr := flag.String("actionStr", "", "Choose an action from convert/cleanupOld/cleanupNew")
+	dataFile := flag.String("data", "", "Enter the path to a data file")
+	action := Convert
+	switch *actionStr {
+	case "convert":
+		action = Convert
+		break
+	case "cleanupOld":
+		action = CleanupOld
+		break
+	case "cleanupNew":
+		action = CleanupNew
+		break
+	default:
+		action = NoAction
+	}
+	flag.Parse()
+
+	for _, handler := range conversionHandlers {
+		log.Printf("Running Conversion handler: %s", handler.description)
+		err := handler.handlerFunc(action, *dataFile)
+		if err != nil {
+			log.Fatalf("An error occured %s", err)
```

Comment:
This is missing content since you do not include the action and the handler which failed.

Is it reasonable to fail after doing some conversions and not others, or should the code keep the old and have a way to fall back (remove any converted output?)

What should the caller (device-steps.sh) do when this fails? reboot? carry on an nothing happened?
@kalyan-nidumolu we should talk about that the high-level failure handling, which is separate from getting useful error messages out for all possible failure cases.

Reply:
Had a discussion with Erik - Change this to log.Errorf and print the error. We should not crash here.

Reply:
Yeah so overall we should log errors (log.Errorf) such as opening files, reading then, json.decode failing, and teating those the same as if the file didn't exist. Hence we need to ignore the timestamp on a file we fail to json.Decode.

Then on those failures we just create a new file with the default values and continue.

A json.Encode should never fail, so that can be log.Fatal. But that is about it.

Reply:
Ok. Thank you

Reply:
Should writing files also be Fatalf?